src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertLinqQueryToForEachTests.cs

```diff
@@ -0,0 +1,2246 @@
+                select n1|];
+        foreach (var b in q)
+        {
+            Console.WriteLine(b);
+        }
+    }
+}
+";
+
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        IEnumerable<int> localFunction()
+        {
+            foreach (int n1 in nums)
+            {
+                foreach (int n2 in nums)
+                {
+                    yield return n1;
+                }
+            }
+        }
+
+        var q = localFunction();
+        foreach (var b in q)
+        {
+            Console.WriteLine(b);
+        }
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task UsageInForEachSameVariableName()
+        {
+            string source = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        var q = [|from int n1 in nums 
+                from int n2 in nums
+                select n1|];
+        foreach(var n1 in q)
+        {
+            Console.WriteLine(n1);
+        }
+    }
+}
+";
+
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        IEnumerable<int> localFunction()
+        {
+            foreach (int n1 in nums)
+            {
+                foreach (int n2 in nums)
+                {
+                    yield return n1;
+                }
+            }
+        }
+
+        var q = localFunction();
+        foreach(var n1 in q)
+        {
+            Console.WriteLine(n1);
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task LinqDefinedInForEach()
+        {
+            string source = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        foreach(var b in [|from int n1 in nums 
+                from int n2 in nums
+                select n1|])
+        {
+            Console.WriteLine(b);
+        }
+    }
+}
+";
+
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                var b = n1;
+                Console.WriteLine(b);
+            }
+        }
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task LinqDefinedInForEachSameVariableName()
+        {
+            string source = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        foreach(var n1 in [|from int n1 in nums 
```

Comment:
❗️ You have a syntax error in the input.

1. The test isn't producing what it should be (`n1` used in `Console.WriteLine` does not refer to what the user intended)
2. The refactoring should not probably have been offered

Reply:
@sharwell, could you please point me to the syntax error? I double-checked by eye and with VS. Still could not find.


---
In reply to: [176527466](https://github.com/dotnet/roslyn/pull/25362#discussion_r176527466) [](ancestors = 176527466)